### ISSUE TITLE
Write Orchestration System Prototype

### ISSUE THREAD

**Issue by jared321** [ECP-Astro/FLASH5#599]
Port the C++/Cuda Orchestration System to Fortran and so that it can run Sedov/Sod with simpleUnsplit on Summit.  Ideally, as a first version, this will include a CudaFortran implementation of the runtime so that we can learn about the low-level aspects of transfers to the GPU and memory management between the CPU and GPU memories.

The kernels will likely be decorated with OpenACC.

**Comment by jared321** [comment]
Hopefully, we will have a reasonably functioning prototype functioning in time for collaborating with Wahib in October.

**Comment by jared321** [comment]
As I work on this, I am including in the standard `!DEV:` comment header (with `TODO` and `FIXME`) to highlight questions and tag code that is meant to be temporary.  

Also, I am labeling code regions that are to be written by the Task Composer with the word `TASK_COMPOSER` so that we can quickly find these regions.

**Comment by jared321** [comment]
@rajeeja I am now able to run a simple simulation that uses a CUDA Fortran file and can communicate with the GPU minimally.  The following can be run on Summit
`./setup Sod -makefile=pgi -auto -2d threadBlockList=True -debug -unit=physics/Hydro/HydroMain/OrchTest -unit=Orchestration/OrchestrationMain/Cuda -nxb=8 -nyb=8 -parfile=test_amr_2d.par --unit=Grid/GridMain/AMR/Amrex +serialio --unit=IO/IOMain/hdf5/serial/AM --index-reorder`

Note that the setup option `threadBlockList=True` is now setting `USEOPENMP=1` in the Makefile for us.

**Comment by jared321** [comment]
@kweide @rajeeja The TestRuntime unit test is now running successfully on Summit using the GPU.  Note that the runtime is programmed in CUDA Fortran while the single loop nest has been kernelized with my best guess at a reasonable OpenACC directive.

```
./setup -makefile=pgi -auto -2d -debug threadBlockList=True unitTest/Orchestration/TestRuntime -nxb=8 -nyb=4 +noio -index-reorder -defines=DEBUG_ORCHESTRATION
```

**Comment by rajeeja** [comment]
@jared321 I"ve PM'ed you summit issues. For site hello, which was working, I now get: 

ERROR Checking Links: Both Orchestration_executeTask.cuf and Orchestration_executeTask.F90 should not be in object directory

**Comment by jared321** [comment]
@rajeeja First, please make certain that you are using the latest commit on the Issue599 branch.

As I have been implementing this work, I have had to change the extensions of some files from F90 to CUF because changes included adding in CUDA Fortran code.  If the setup script is complaining about this, then it is still finding both versions scattered throughout the Orchestration unit folder.

While this should not be happening, I wonder if git has left the F90 version in your repository despite having made that version untracked in the repository.  If you do a `git status` from the root of this folder, hopefully you will see which version of the file is untracked (hopefully the F90 version).  Then you can delete this and retry the setup.

I see that there is an F90 version of this file in the CpuTest subfolder.  However, if your setup is for the Cuda implementation, this file should be ignored by the setup script.

**Comment by rajeeja** [comment]
I have the latest version. My setup script is for CpuTest, so it should ignore Orchestration_executeTask.cuf, but it doesn't. 

./setup Sod -site=hello -auto -2d threadBlockList=True  -debug -unit=physics/Hydro/HydroMain/OrchTest -unit=Orchestration/OrchestrationMain/CpuTest -nxb=8 -nyb=8 -parfile=test_amr_2d.par --unit=Grid/GridMain/AMR/Amrex +serialio --unit=IO/IOMain/hdf5/serial/AM --index-reorder -objdir=hello_sod_orch

**Comment by jared321** [comment]
Don't use the CpuTest implementation anymore as I am not maintaining it right now and it will disappear sooner rather than later.  Use the Cuda implementation.

**Comment by rajeeja** [comment]
gotcha! In that case Cuda implementation should work for Cpu also (meaning machines without GPU-Cuda)?
        1
Fatal Error: Can't open module file ‘or_cudaptrs.mod’ for reading at (1): No such file or directory
compilation terminated.
Makefile:117: recipe for target 'Orchestration_tileList.mod' failed
make: *** [Orchestration_tileList.mod] Error 1
make: *** Waiting for unfinished jobs....
io_xfer_tree_data.F90:95:42:

**Comment by jared321** [comment]
Nope.  The TestRuntime unittest and simulations setup with the OrchTest Hydro implementation are hardwired to GPUs right now and the OrchTest code should fail because the loop nests have not been kernelized yet with OpenACC.

**Comment by jared321** [comment]
@rajeeja If you want to get things running, try the following on Summit:

```
./setup -makefile=pgi -auto -2d -debug threadBlockList=True unitTest/Orchestration/TestRuntime -nxb=8 -nyb=4 +noio -index-reorder -defines=DEBUG_ORCHESTRATION
```
